pact/brackets-review.pact

```diff
@@ -0,0 +1,504 @@
+    winner:string
+    entryFee:integer)
+
+  (deftable bracket-betting-table:{bracket-betting})
+
+  (defschema users
+    username:string
+    user-guard:guard
+    balance:integer
+    bb-games:list
+    bb-admins:list
+    eb-games:list
+    eb-admins:list
+    games-won:list)
+
+  (deftable users-table:{users})
+
+  (defcap BRACKET-ADMIN-BB (admin-key:string bracket-name:string)
+    (with-read bracket-betting-table bracket-name { "admin":= admin-key-db }
+      (enforce (= admin-key admin-key-db) "you are not the bracket admin")
+    )
+  )
+
+  (defcap BRACKET-ADMIN-EB (admin-key:string bracket-name:string)
+    (with-read empty-bracket-table bracket-name { "admin":= admin-key-db }
+      (enforce (= admin-key admin-key-db) "you are not the bracket admin")
+    )
+  )
+
+  (defcap USER-CAN-TRANSFER (from:string amount:integer)
+    (with-read users-table from { "balance" := from-bal }
+      (enforce (>= from-bal amount) "Insufficient Funds")
+    )
+  )
+
+  (defcap EB-CONTRACT-CAN-TRANSFER (bracket-name:string amount:integer)
+    (with-read empty-bracket-table bracket-name { "moneyPool" := from-bal }
+      (enforce (>= from-bal amount) "Insufficient Funds")
+    )
+  )
+
+  (defcap BB-CONTRACT-CAN-TRANSFER (bracket-name:string amount:integer)
+    (with-read bracket-betting-table bracket-name { "moneyPool" := from-bal }
+      (enforce (>= from-bal amount) "Insufficient Funds")
+    )
+  )
+
+  (defcap IS-REGISTERED-USER (user-key:string)
+    (let ((users (keys users-table)))
+          ;should be user-id
+        (enforce (contains user-key users) "you are not a registered user")
+    )
+  )
+
+  ;needs to be passed as arg to function (the guard)
+  ;user-key is both id and documents user from payload
+  ;usually its keyset in the arguments
+  (defun init-user (user-key:string username:string)
+    ;user id
+    (insert users-table user-key {
+      "username": username,
+      "balance": 100,
+      "user-guard": (read-keyset user-key),
+      "bb-games": [],
+      "bb-admins": [],
+      "eb-games": [],
+      "eb-admins": [],
+      "games-won": []
+    })
+  )
+
+  (defun get-user-info (user-key:string)
+    (with-capability (IS-REGISTERED-USER user-key)
+        (with-read users-table user-key {
+          "username":=username,
+          "bb-games":=bb-games,
+          "bb-admins":=bb-admins,
+          "eb-games":=eb-games,
+          "eb-admins":=eb-admins,
+          "balance":= balance,
+          "games-won":= games-won}
+          [username, bb-games, bb-admins, eb-games, eb-admins, balance, games-won]
+        )
+    )
+  )
+
+  (defun get-all-users ()
+    ;;anyone can call this
+    (keys users-table)
+  )
+
+  (defun init-empty-bracket
+    (bracket-name:string bracket:list number-players:integer entry-fee:integer)
+    "initiate a new bracket"
+    ;sender is id not key...
+    (let ((admin-key (at "sender" (chain-data))))
+    ;is already a user of our systems
+    (with-capability (IS-REGISTERED-USER admin-key)
```

Comment:
Here it makes sense to want to check that the admin exists (as opposed to above in `get-user-info`). However, this isn't a capability really. Capabilities are not like Ethereum modifiers, they are supposed to represent some unforgeable truth, not just some assertion. See my comment on `IS-REGISTERED-USER`.